Only compute partial paths from references

hendrikvanantwerpen opened the pull request:
Partial paths are now computed from references only, instead of from any push node. This aligns with the existing behavior for scopes, where only exported scopes are used as start nodes.

hendrikvanantwerpen reviewed (commented):
We start building partial paths from endpoints, but visit _all_ paths starting from these nodes. Would it make sense to guard [visit](https://github.com/github/stack-graphs/blob/main/stack-graphs/src/partial.rs#L2457) with a check for `path.is_as_complete_as_possible(graph)`?

dcreager reviewed (approved):
> Would it make sense to guard [visit(...)](https://github.com/github/stack-graphs/blob/main/stack-graphs/src/partial.rs#L2457) with a check for `path.is_as_complete_as_possible(graph)` instead of leaving that to the caller?

Yes, that's a good catch!  Though I think the call to [`extend_from_file`](https://github.com/github/stack-graphs/blob/main/stack-graphs/src/partial.rs#L2456) also needs to be guarded.  That's the bit that adds the path's extensions to the work queue; skipping that step would cut off the search at that point.  (And note that in the [old internal implementation](https://github.com/github/tree-sitter-code-nav/blob/master/lib/src/pathfind.rs#L194-L195), this step is guarded both by `is_complete_as_possible` _and_ `is_productive`.)

hendrikvanantwerpen commented:
> Though I think the call to [extend_from_file](https://github.com/github/stack-graphs/blob/main/stack-graphs/src/partial.rs#L2456) also needs to be guarded. That's the bit that adds the path's extensions to the work queue; skipping that step would cut off the search at that point.

I am not sure. We _do_ want to extend paths that are _not_ as complete as possible. We could cut off the search if a path is as complete as possible, but I think this would cause problems in alias-like situations, where you have complete paths linked by an edge from the definition to another reference. Cutting of would mean the longer path to the aliased definition is not visited. An alternative solution to that problem could be to also generate paths from definitions to (not necessarily matching) references.

> (And note that in the [old internal implementation](https://github.com/github/tree-sitter-code-nav/blob/master/lib/src/pathfind.rs#L194-L195), this step is guarded both by is_complete_as_possible and is_productive.)

That's a good point, I'll check both!

dcreager commented:
> I am not sure. We _do_ want to extend paths that are _not_ as complete as possible. We could cut off the search if a path is as complete as possible, but I think this would cause problems in alias-like situations, where you have complete paths linked by an edge from the definition to another reference. Cutting of would mean the longer path to the aliased definition is not visited. An alternative solution to that problem could be to also generate paths from definitions to (not necessarily matching) references.

Ah I misread the link I sent!  I thought that my link showed that we _were already_ cutting off the paths that were complete-as-possible and productive.  But I'm wrong, that's just what limits whether it's included in the result set, just like we're currently doing here in the new open-source crate.

We _do_ have a [`path_should_extend`](https://github.com/github/tree-sitter-code-nav/blob/master/lib/src/pathfind.rs#L399-L407) check in the old implementation, but it's very simple: we cut off paths at the root node and at any endpoint node.  We should make sure that cutoff logic exists here in the open-source crate too.

So, given this, I think it's still better to have the complete-as-possible/productive calls _in_ the visitor (like they are now), instead of moving them into `find_all_partial_paths_in_file` as you suggested.  There wouldn't be an efficiency gain, since `visit` function has a (static) generic type, so the call will already get monomorphized and inlined into `find_all_partial_paths_in_file` with the way that it's written now.

hendrikvanantwerpen commented:
> We do have a [path_should_extend](https://github.com/github/tree-sitter-code-nav/blob/master/lib/src/pathfind.rs#L399-L407) check in the old implementation, but it's very simple: we cut off paths at the root node and at any endpoint node. We should make sure that cutoff logic exists here in the open-source crate too.

The endpoint check is what I was arguing _against_. I think we'd miss aliases because of that. Stopping at the root node seems fine.

> So, given this, I think it's still better to have the complete-as-possible/productive calls in the visitor (like they are now), instead of moving them into find_all_partial_paths_in_file as you suggested. There wouldn't be an efficiency gain, ...

My reasoning for putting the check inside `find_all_partial_paths_in_file` was not for efficiency reasons, but for ergonomics. Would callers ever want anything but that check in the visitor, especially since this already a fairly high-level convenience function? And it might be easy to forget that check, or not realize the necessity.

dcreager commented:
> The endpoint check is what I was arguing _against_. I think we'd miss aliases because of that. Stopping at the root node seems fine.

Ah, gotcha.  Are you considering def/ref nodes "endpoints"?  I think in the old implementation "endpoint" only refers to endpoint _scope_ nodes.  And in that interpretation, we'd be okay, because any endpoint that we'd cut off a path at is _also_ a valid starting node for outgoing paths.  And at _stitching_ time, we'd join those paths together to handle the alias.

But you're right that we wouldn't want to cut off a path once it reached a definition node, because there might be outgoing edges that we want to follow, and a definition node is not a valid starting point for the partial-path-finding algorithm.

hendrikvanantwerpen commented:
I realize now this is a little more subtle than I expected.

Let's start with the goals:
1. The set of all partial paths visited should be enough to stitch all complete paths (ie references to definitions).
2. Ideally we visit as few paths as possible for that.

Since we start at root, exported scopes and references, we can stop extending any non empty path ending in any of these nodes.

To ensure everything can be reached, we should visit non empty paths from root, exported scopes and references to root, exported scopes, references and definitions. The last one because I assume that during stitching we only consider the end nodes of the paths and don't look at the intermediate nodes anymore. This is more liberal than the `is_complete_as_possible` check.

The effect of this approach is that some visited paths will overlap. Those to a definition that are later extended to a reference.

An alternative could be to visit only paths to definitions, not references, and additionally generate partial paths from definitions to references.

Thoughts, @dcreager?

dcreager commented:
> I realize now this is a little more subtle than I expected.

This is a good summary of the subtlety, and I think the goals that you call out are spot on.

As a way of trying to minimize the subtlety, what do you think about making the sets of valid starting and ending nodes be identical?  They would be: root, exported/endpoint nodes, references, and definitions.  We would accept any partial path that started and ended at _any_ of those kinds of nodes, instead of having different combinations of valid starting/ending nodes, based on what we think we'll need at stitching time.  That might make us generate partial paths that could never be useful at stitching time, but I imagine that it would be a small number, relative to the number of "real" paths we'll find, and I think it could greatly simplify the logic.  (And I think it would subsume your suggestion + your alternative.)

hendrikvanantwerpen commented:
Excellent idea! I was already coming up with cases I hadn't considered the moment I posted that comment. Your suggestions would make us visit perhaps a few more path, but give us the confidence we certainly have what we need. If that turns out to be too much still, we'll revisit.

hendrikvanantwerpen commented:
@dcreager Your proposal is good for deciding which paths to visit, but I don't see the corresponding cut-off criterium. Do you think we'll be okay without one altogether? After all, cycle detection and the regular path extension logic will rule out incompatible combinations.

hendrikvanantwerpen commented:
Ah, yes, that would work. Okay, I'll give that a go.

hendrikvanantwerpen commented:
@dcreager Something funny happens when I end changed the logic to what we discussed. Take the following step in stiching:

```
--> Candidate path [main.py(6) reference foo] -> [main.py(8) reference a] <a.foo> ()
      Search for start node [main.py(8) reference a]
        Found path <%1> () [main.py(8) reference a] -> [root] <a,%1> ()
    Extend [main.py(6) reference foo] -> [main.py(8) reference a] <a.foo> ()
      with <%1> () [main.py(8) reference a] -> [root] <a,%1> ()
        is [main.py(6) reference foo] -> [root] <aa.foo> ()
                                                 ^^ <--- duplicated symbol
```

If the shared node between the two partial paths manipulates the stack, that seems to be applied twice.

hendrikvanantwerpen commented:
@dcreager I backtracked on this approach and opted for something more expedient. I changed the starting nodes to only include references (not all push nodes), and cut-off path extension at root. I also added the visitor checks _inside_ the function, added documentation that clarifies thata whatever this function visits is sufficient for stitching complete paths, and removed the checks from all use sites. This will give us more flexibility in the future _if_ we want to revisit this implementation, but seems to be good enough for now---memory usage and runtime are still improved compared to what it used to be.

dcreager commented on stack-graphs/CHANGELOG.md:
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
+
+### Changed
+
+- The amount of work done in `PartialPaths::find_all_partial_paths_in_file` is reduced. The documentation now makes it clear that the function itself is responsible for returning a set that is big enough to cover all complete paths. Callers should not have to filter this set further anymore, and any visitors filtering with `PartialPath::is_complete_as_possible` or `PartialPath::is_productive` sho ...
```
spelling nit: optimizations

dcreager reviewed (approved):
:+1: I like expedient!
